### PR TITLE
Change: [Linkgraph] Allow job threads to be aborted early when clearing schedule

### PR DESCRIPTION
**Pull request by JGRennison**
## Motivation / Problem

Long running link graph jobs running in separate threads continue running until completion when the game state is cleared due to one of: abandoning game, loading replacement game, exiting game.

This causes the game and user interface to be unnecessarily blocked by joining the job thread when the result will be discarded.

## Description

When link graph jobs are cleared due to abandoning the game or exiting, flag the job as aborted.

The link graph job running in a separate thread checks the aborted flag periodically and terminates processing early if set.

This reduces the delay at game abandon or exit if a long-running job would otherwise still be running.

## Limitations

N/A

## Checklist for review

PR justification: https://github.com/OpenTTD/OpenTTD/pull/7081#pullrequestreview-557094105

### CODE REVIEW

File: src/linkgraph/linkgraphschedule.cpp [hunk 1]
```diff
@@ -119,7 +120,7 @@ void LinkGraphSchedule::SpawnAll()
 /* static */ void LinkGraphSchedule::Clear()
 {
 	for (JobList::iterator i(instance.running.begin()); i != instance.running.end(); ++i) {
-		(*i)->JoinThread();
+		(*i)->AbortJob();
```
Comment (michicc) [inline src/linkgraph/linkgraphschedule.cpp:123]: I don't think you can loose the `JoinThread` here. It is required to call either `join` or `detach` on a `std::thread` before destruction.

Comment (JGRennison) [reply src/linkgraph/linkgraphschedule.cpp:123]: JoinThread is called in ~LinkGraphJob, it's not necessary to call it twice.

Comment (michicc) [reply src/linkgraph/linkgraphschedule.cpp:123]: Ah, sorry, somehow missed that.

**Comment by TrueBrain** [comment]
I did a quick test of this PR, where I add a `CSleep(30000)` in the `LinkGraphSchedule::Run()`, as with the other PR.

This does not work for me as advertised; "Abandon Game", "Exit Game", and "Load Game" still hangs for 30 seconds before the game continues.

Looking at the code, that somewhat makes sense, so I moved the `CSleep` into `FlowMapper::Run()`. Still this hangs my client for 30 seconds in all cases.

Am I right to assume this happens because this PR assumes only the MCF can take a lot of time? And not for example the flowmapper? And even for the MCF you assume a single iteration is not the problem, but the amount of iterations it is doing?

To be clear, not saying you should fix my cases, just trying to understand what this does and doesn't fix. Mainly as the PR description suggests a broader scope than the code does ;)
I was also wondering why you don't just terminate the thread, instead of asking the thread to terminate as quickly as it can? (again, honest question, I know very little about this code, so all I can do is ask stupid questions :D)

Tnx!

**Comment by JGRennison** [comment]
A single large sleep is by nature not interruptible by this sort of mechanism.

> I was also wondering why you don't just terminate the thread, instead of asking the thread to terminate as quickly as it can? (again, honest question, I know very little about this code, so all I can do is ask stupid questions :D)

This is highly unportable and has a high risk of putting the whole process in a bad state.

For example, if the thread is holding a lock or some other resource when you sent it a kill signal, that lock could be left locked, permanently wedging the whole process. You would also have a high risk of the LinkGraphJob destructor exploding when you try to clean up from the main thread side, unless you decided to leak it on purpose. There are ways around this like pthread_cleanup but these are extremely fragile. I would not recommend such measures.

> Am I right to assume this happens because this PR assumes only the MCF can take a lot of time? And not for example the flowmapper? And even for the MCF you assume a single iteration is not the problem, but the amount of iterations it is doing?

The first MCF pass is by far the most expensive phase.
There is a trade off between the numbers of abort checks and the delay before the next abort check is reached.
In general there is no "correct" number of abort check points, but the checks as included gave acceptable latency in my own testing, such that more fine grained checking was unnecessary.
The testing was done several years ago, but the link graph has not meaningfully changed since then.

**Review by TrueBrain — APPROVED** [review]
Perfectly clear explanation :D Thank you!
For future reference, that would have been perfect to mention in "Limitations" 😄 But I fully understand it now, so this seems more than fine by me!

tldr; doesn't solve all cases, but it does solve the most important ones. If we ever find others that are lagging, we can deal with those when they become apparent :D
